core/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/AvgAggregator.java

```diff
@@ -76,15 +78,31 @@ public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
             public void collect(int doc, long bucket) throws IOException {
                 counts = bigArrays.grow(counts, bucket + 1);
                 sums = bigArrays.grow(sums, bucket + 1);
+                compensations = bigArrays.grow(compensations, bucket + 1);
 
                 if (values.advanceExact(doc)) {
                     final int valueCount = values.docValueCount();
                     counts.increment(bucket, valueCount);
-                    double sum = 0;
+                    // Compute the sum of double values with Kahan summation algorithm which is more
+                    // accurate than naive summation.
+                    double sum = sums.get(bucket);
+                    double compensation = compensations.get(bucket);
+
                     for (int i = 0; i < valueCount; i++) {
-                        sum += values.nextValue();
+                        double value = values.nextValue();
+                        if (Double.isNaN(value) || Double.isInfinite(value)) {
+                            sum += value;
+                            if (Double.isNaN(sum))
+                                break;
+                        } else if (Double.isFinite(sum)) {
```

Comment:
this is always true I think? So we could make it an `else` block?

Reply:
The check statement is needed. For example when summing up ```[Double.POSITIVE_INFINITY, 1, 2, 3]```, if no this check, we'll got ```NaN```, and we expect ``` Double.POSITIVE_INFINITY``` here.